### PR TITLE
Add backup and restore scripts for Linux/MacOS

### PR DESCRIPTION
**Pull request by BClark09**
Related to #499 

Signed-off-by: Ben Clark <ben@benjyc.uk>

### CODE REVIEW

**Comment by BClark09** [comment]
That's my best stab at it for now, would appreciate reviews, suggestions and testing @ThomDietrich and @whopperg :)

**Comment by BClark09** [comment]
image

**Comment by thopiekar** [comment]
Like it! I previously also added something like a backup mechanism for QNAP NASs. As soon as this is ready I'll adopt it.

**Comment by BClark09** [comment]
Thanks @thopiekar, looking at the folder structure it for QNAP NASs it should work fine, have you already tested it for it?

@kaikreuzer, could you test these scripts on MacOS?

**Comment by ThomDietrich** [comment]
Hey @BClark09, I went through the code and everything looks great.

The only thing I still don't quite like is "Usage: ./runtime/bin/backup [filename]". Feels unnatural to demand a relative path for execution of a script. As a user I'd like to run the script with an absolute path, from inside the bin folder.

Furthermore I'm wondering if we should add symlinks to these scripts (and others) to `/usr/bin` and `/usr/sbin` in the https://github.com/openhab/openhab-linuxpkg project.

**Comment by thopiekar** [comment]
@BClark09 No, I haven't but implemented a simple function to my service script to make backups of the whole distribution (only working when openHAB is offline, of course.).
So yeah, nice to have a commonized script/function now!

**Comment by BClark09** [comment]
@ThomDietrich ,

> The only thing I still don't quite like is "Usage: ./runtime/bin/backup [filename]". Feels unnatural to demand a relative path for execution of a script/program

This is only necessary if the OPENHAB_* environment paths are unknown. It'll work anywhere otherwise. I'd rather not try to work out where the appropriate path is from the file location, since this starts to break down with symlinks.

> Furthermore I'm wondering if we should add symlinks to these scripts to /usr/bin and /usr/sbin in the https://github.com/openhab/openhab-linuxpkg project.

Actually, I was considering adding /usr/bin/openhab soon to add the commands:

```
openhab console
openhab backup $@
openhab restore $@
openhab showlogs
```

thoughts?

**Comment by thopiekar** [comment]
I agree with @ThomDietrich. It would be nice, if the filename would be optional. This way people can decide whether they want to create a archive inside the distro or eg. outside.

**Comment by kaikreuzer** [comment]
> @kaikreuzer, could you test these scripts on MacOS?

Tested, not working. I get the message (both on backup and restore)
```
cp: the -R and -r options may not be specified together.
cp: the -R and -r options may not be specified together.
```
and the backup only contains empty conf and userdata folders.

Regarding the message on restore:

> Any existing files with the same name will be replaced, but this will NOT delete existing files that have no replacement.

Why not? If I restore a backup, I would want the setup of that moment and not some mix of current stuff and restored stuff. Or do I misinterpret the notice?

**Comment by thopiekar** [comment]
@kaikreuzer That's why I make a full copy of the whole distro and then if something goes wrong I remove everything and restore all from the backup.
Just like I would expect it from a 'snapshot' of a known working state.

**Comment by BClark09** [comment]
Alright will rework. Not sure about that -R, -r error though. It only uses one.

**Comment by kaikreuzer** [comment]
Checking the man page it says:

```
     -a    Same as -pPR options. Preserves structure and attributes of files but not directory structure.
```
so by specifying `-a` you implicitly have a `-R` as it seems.

**Comment by kaikreuzer** [comment]
... and indeed, when I remove the `r`, the backup script seems to work nicely 👍

**Comment by BClark09** [comment]
Alright, made a couple of changes that should mean that the script can be called from anywhere, and now should work on MacOS (Thanks @kaikreuzer!)

> I would want the setup of that moment and not some mix of current stuff and restored stuff. Or do I misinterpret the notice?

You didn't misinterpret, but I understand this is probably the desired choice. How best to proceed with that then? Delete all but the specific `userdata/etc/` files and then the backup on top of the somewhat empty directory?

**Comment by BClark09** [comment]
Hi all, I've made some adjustments so that it completely replaces userdata and conf, getting rid of anything that wasn't in the backup.

Please could you check and test these new changes, as we need to eliminate anything that could be potentially dangerous.

**Comment by bdleedy** [comment]
I'm writing the Windows version of this and I'm noticing that you put the backup in runtime\bin\backups. If you would then use upgrade that backup would then be deleted. Is this intended? Is there a plan to modify upgrade to handle retaining the backup?

**Comment by BClark09** [comment]
Hi @bdleedy, the backup should be stored in openhab_home/backups by default. Is that not the case here?  The line:

```
 WorkingDir="$(cd "$(dirname "$0")" && cd ../.. && pwd -P)"
```

should point two directories up from where the backup script is.

**Comment by bdleedy** [comment]
NM. Duh. Apparently too much code floating around in my head yesterday. Thanks.

**Comment by BClark09** [comment]
Alright, I think I've finished with this. @ThomDietrich, @kaikreuzer and @thopiekar, I've taken all comments on board. Please let me know how you get on with testing and reviewing the two scripts. I've tested on a debian virtual machine, but please try to break the scripts as much as possible.

File: distributions/openhab/src/main/resources/bin/restore [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  }
+fi
+
+echo "Deleting old conf folder..."
+rm -rf "${OPENHAB_CONF:?}/"*
+
+
+## Restore configuration
+echo "Restoring openHAB with backup configuration..."
+command cp -af "$TempDir/conf/"*     "${OPENHAB_CONF:?}/" || {
```
Comment (ThomDietrich) [inline distributions/openhab/src/main/resources/bin/restore]: what's the `command` needed for?

Comment (BClark09) [inline distributions/openhab/src/main/resources/bin/restore]: Some distributions use an alias for cp which includes the -i flag. 

https://serverfault.com/questions/119869/turning-off-cp-copy-commands-interactive-mode-cp-overwrite

**Comment by kaikreuzer** [comment]
@BClark09 I have tested it on Mac, seems to run smoothly! 👍

**Comment by kaikreuzer** [comment]
Any further comments/tests by the others? Or do you think we are good to merge?

**Comment by ThomDietrich** [comment]
I did not get around doing any tests (my time is a bit limited atm) but I went through the whole code and it LGTM.

**Comment by BClark09** [comment]
Hi @ThomDietrich and @kaikreuzer, would you be able to test this once more? I cleaned up and used the scripts on my own home today so these should be fine to use. I've tried to harden the script as much as possible but please try to break it if you can :)

**Comment by ThomDietrich** [comment]
LGTM. Great work @BClark09 !

**Comment by whopperg** [comment]
TOP!!! @BClark09

**Comment by kaikreuzer** [comment]
Thanks @BClark09, successfully tested it on Mac as well!

**Comment by BClark09** [comment]
Awesome, thanks everyone! Shall I put a post up on the community forums about this?

**Pull request merged**